cli/azd/test/functional/env_test.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cli_test
+
+import (
+	"context"
+	"encoding/json"
+	"testing"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/contracts"
+	"github.com/azure/azure-dev/cli/azd/test/azdcli"
+	"github.com/stretchr/testify/require"
+)
+
+// Verifies azd env commands that manage environments.
+func Test_CLI_Env_Management(t *testing.T) {
+	ctx, cancel := newTestContext(t)
+	defer cancel()
+
+	dir := tempDirWithDiagnostics(t)
+	t.Logf("DIR: %s", dir)
+
+	cli := azdcli.NewCLI(t)
+	cli.WorkingDirectory = dir
+
+	err := copySample(dir, "storage")
+	require.NoError(t, err, "failed expanding sample")
+
+	// Create one environment via interactive prompt
+	envName := randomEnvName()
+	envNew(ctx, t, cli, envName, true)
+
+	// Verify env list is updated
+	environmentList := envList(ctx, t, cli)
+	require.Len(t, environmentList, 1)
+	requireIsDefault(t, environmentList, envName)
+
+	// Create one environment via flags
+	envName2 := randomEnvName()
+	envNew(ctx, t, cli, envName2, false)
+
+	// Verify env list is updated, and with new default set
+	environmentList = envList(ctx, t, cli)
+	require.Len(t, environmentList, 2)
+	requireIsDefault(t, environmentList, envName2)
+
+	// Select old environment
+	envSelect(ctx, t, cli, envName)
+
+	// Verify env list has new default set
+	environmentList = envList(ctx, t, cli)
+	require.Len(t, environmentList, 2)
+	requireIsDefault(t, environmentList, envName)
+}
+
+// Verifies azd env commands that manage environment values.
+func Test_CLI_Env_Values_Management(t *testing.T) {
+	ctx, cancel := newTestContext(t)
+	defer cancel()
+
+	dir := tempDirWithDiagnostics(t)
+	t.Logf("DIR: %s", dir)
+
+	cli := azdcli.NewCLI(t)
+	cli.WorkingDirectory = dir
+
+	err := copySample(dir, "storage")
+	require.NoError(t, err, "failed expanding sample")
+
+	// Create one environment
+	envName := randomEnvName()
+	envNew(ctx, t, cli, envName, false)
```

Comment:
Is this enough to cover all the scenarios?
I wish we could have unit tests as well, so we don't only rely on functional tests.

Since the code we are adding is within the action.Run() and build, we should be able to test the action directly and have coverage here:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1971865&_a=summary&tab=ms.vss-codecoverage-web.code-coverage-details&view=codecoverage-tab

Reply:
I wouldn't consider substituting these tests for unit tests. With the way these tests are written, what we're testing is really the integration between commands, which I think is extremely valuable. It does also cover almost all executing code in the action given the minimal branching paths by coincidence.

I would consider adding tests on the action if there were more interesting things to validate (specific outputs, different input conditions). I would also consider testing on the action if the end-to-end test were simply too slow, and we gain value by mocking out pieces of the action and testing that in isolation. (For example, mocking out ARM).